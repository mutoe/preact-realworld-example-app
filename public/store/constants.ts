export const UPDATE_USER = 'UPDATE_USER';
export const SET_ERRORS = 'SET_ERRORS';
export const CLEAN_ERRORS = 'CLEAN_ERRORS';

export const DEFAULT_AVATAR = 'https://static.productionready.io/images/smiley-cyrus.jpg';
