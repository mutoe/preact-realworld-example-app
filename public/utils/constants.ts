export const DEFAULT_AVATAR = 'https://static.realworld.io/images/smiley-cyrus.jpg';
