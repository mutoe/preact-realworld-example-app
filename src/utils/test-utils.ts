import mockjs, { Random } from 'mockjs'
import { ReactWrapper } from 'enzyme'

export function generateProfile (): Profile {
  return mockjs.mock({
    username: Random.name(),
    bio: Random.sentence(),
    image: Random.image(),
    following: Random.boolean(),
  })
}

export function generateArticles(): Article
export function generateArticles(count: number): Article[]
export function generateArticles (count = 1): Article | Article[] {
  const title = Random.title()
  const slug = `${title.replace(/[ -]/g, '-')}-${Number(new Date()).toString(36)}`
  return mockjs.mock({
    [`articles|${count}`]: [
      {
        title,
        slug,
        body: Random.paragraph(),
        createdAt: new Date(Random.date()).toISOString(),
        updatedAt: new Date(Random.date()).toISOString(),
        'tagList|3': [() => Random.word()],
        description: Random.sentence(),
        author: () => generateProfile(),
        favorited: Random.boolean(),
        'favoritesCount|0-200': 200,
      },
    ],
  }).articles
}

export const setInputValue = (wrapper: ReactWrapper<any>, selector: string, value: string) => {
  wrapper.find(selector).getDOMNode<HTMLInputElement>().value = value
  wrapper.find(selector).simulate('input')
}

export const getInputValue = (wrapper: ReactWrapper<any>, selector: string): string => {
  return wrapper.find(selector).getDOMNode<HTMLInputElement>().value
}

export function generateComments(): ArticleComment
export function generateComments(count: number): ArticleComment[]
export function generateComments (count = 1): ArticleComment | ArticleComment[] {
  return mockjs.mock({
    [`comments|${count}`]: [
      {
        id: mockjs.Random.integer(0, 100),
        body: '@paragraph',
        createdAt: new Date(Random.date()).toISOString(),
        updatedAt: new Date(Random.date()).toISOString(),
        author: () => generateProfile(),
      },
    ],
  }).comments
}
