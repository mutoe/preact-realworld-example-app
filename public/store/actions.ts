import { postLogin, PostLoginForm } from '../services';
import { SET_ERRORS, UPDATE_USER } from './constants';

export const login = async (form: PostLoginForm): Promise<Action> => {
    try {
        const user = await postLogin(form);
        return {
            type: UPDATE_USER,
            user,
        };
    } catch (e) {
        return {
            type: SET_ERRORS,
            errors: e.errors,
        };
    }
};
