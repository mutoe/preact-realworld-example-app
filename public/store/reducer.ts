import { CLEAN_ERRORS, SET_ERRORS, UPDATE_USER } from './constants';
import { request } from '../services';

const reducer = (state: RootState, action: Action): RootState => {
    switch (action.type) {
        case SET_ERRORS: {
            return { ...state, errors: action.errors };
        }
        case CLEAN_ERRORS: {
            return { ...state, errors: {} };
        }

        case UPDATE_USER: {
            let user = action.user;

            if (!user) {
                // logout
                global.localStorage.removeItem('user');
                request.options.headers.Authorization = '';
                return { ...state, user: undefined };
            }

            // login
            user = Object.assign({}, state.user, user) as User;
            global.localStorage.setItem('user', JSON.stringify(user));
            request.options.headers.Authorization = `Token ${user.token}`;
            return { ...state, user };
        }

        default:
            return state;
    }
};

export default reducer;
