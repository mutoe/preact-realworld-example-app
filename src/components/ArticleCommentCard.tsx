import { h } from 'preact';

import { apiDeleteComment } from '../services/api/comments';
import useStore from '../store';
import { DEFAULT_AVATAR } from '../utils/constants';
import { dateFormatter } from '../utils/dateFormatter';

interface ArticleCommentCardProps {
	articleSlug: string;
	comment: ArticleComment;
	onDelete: (commentId: number) => void;
}

export default function ArticleCommentCard(props: ArticleCommentCardProps) {
	const { comment } = props;
	const user = useStore(state => state.user);

	const onDelete = async () => {
		await apiDeleteComment(props.articleSlug, comment.id);
		props.onDelete(comment.id);
	};

	return (
		<div class="card">
			<div class="card-block">
				<p class="card-text">{comment.body}</p>
			</div>
			<div class="card-footer">
				<a href={`/@${comment.author.username}`} class="comment-author">
					<img src={comment.author.image || DEFAULT_AVATAR} class="comment-author-img" />
				</a>
				&nbsp;
				<a href={`/@${comment.author.username}`} class="comment-author">
					{comment.author.username}
				</a>
				<span class="date-posted">{dateFormatter(comment.createdAt)}</span>
				<span class="mod-options">
					{user?.username === comment.author.username && <i class="ion-trash-a" onClick={onDelete} />}
				</span>
			</div>
		</div>
	);
}
