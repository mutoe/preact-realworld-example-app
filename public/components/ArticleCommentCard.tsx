import { apiDeleteComment } from '../services/api/comments';
import { useStore } from '../store';
import { DEFAULT_AVATAR } from '../utils/constants';
import { dateFormatter } from '../utils/dateFormatter';

interface ArticleCommentCardProps {
	articleSlug: string;
	comment: ArticleComment;
	onDelete: () => void;
}

export function ArticleCommentCard(props: ArticleCommentCardProps) {
	const { comment } = props;
	const user = useStore(state => state.user);

	const onDelete = async () => {
		await apiDeleteComment(props.articleSlug, comment.id);
		props.onDelete();
	};

	return (
		<div class="card">
			<div class="card-block">
				<p class="card-text">{comment.body}</p>
			</div>
			<div class="card-footer">
				<a href={`/@${comment.author.username}`} class="comment-author">
					<img src={comment.author.image || DEFAULT_AVATAR} class="comment-author-img" alt="User's profile picture" />
				</a>{' '}
				<a href={`/@${comment.author.username}`} class="comment-author">
					{comment.author.username}
				</a>
				<span class="date-posted">{dateFormatter(comment.createdAt)}</span>
				<span class="mod-options">
					{user?.username === comment.author.username && (
						<i class="ion-trash-a" onClick={onDelete} aria-label="Delete comment" />
					)}
				</span>
			</div>
		</div>
	);
}
