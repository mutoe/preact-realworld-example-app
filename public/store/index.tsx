import { createContext, h, VNode } from 'preact';
import parseStorageGet from '../utils/parse-storage-get';
import { useContext, useEffect, useReducer } from 'preact/hooks';
import reducer from './reducer';
import { UPDATE_USER } from './constants';
import { request } from '../services';
import { route } from 'preact-router';

const initialRootState: RootState = {
    user: parseStorageGet('user') || null,
    errors: {},
};

type RootContextProps = [RootState, (action: Action) => void];

export const RootContext = createContext<RootContextProps>({} as RootContextProps);

export const RootProvider = (props: { children: VNode | VNode[] }) => {
    const [state, dispatch] = useReducer(reducer, initialRootState);

    request.options.responseInterceptor = (response) => {
        if (response.status === 401) {
            dispatch({ type: UPDATE_USER });
            route('/login');
        }
    };

    useEffect(() => {
        dispatch({ type: UPDATE_USER, user: state.user });
    }, []);

    return <RootContext.Provider value={[state, dispatch]}>{props.children}</RootContext.Provider>;
};

export const useRootState = () => useContext(RootContext);
