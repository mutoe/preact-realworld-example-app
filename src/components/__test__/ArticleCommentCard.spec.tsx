import render from 'preact-render-to-string'
import ArticleCommentCard from '../ArticleCommentCard'
import { h } from 'preact'
import { shallow } from 'enzyme'
import { generateProfile, generateComments } from '../../utils/test-utils'
import { useRootState } from '../../store'

jest.mock('../../store')

const useRootStateMock = useRootState as jest.Mock

const loggedUser = generateProfile()

beforeEach(() => {
  useRootStateMock.mockReturnValue([{ user: loggedUser }, jest.fn()])
})

afterEach(jest.clearAllMocks)

describe('# Article comment card', function () {
  it('should render correctly', function () {
    const comment: ArticleComment = {
      id: 1,
      body: 'body.',
      createdAt: '2009-02-19T00:00:00.000Z',
      updatedAt: '2008-05-19T00:00:00.000Z',
      author: {
        username: 'Shirley Davis',
        bio: 'bio.',
        image: 'http://dummyimage.com/336x280',
        following: true,
      },
    }
    const output = render(<ArticleCommentCard comment={comment} />)

    expect(output).toMatchSnapshot()
  })

  it('should not display trash icon in other\'s comment', async function () {
    const comment = generateComments()
    const wrapper = shallow(<ArticleCommentCard comment={comment} />)

    expect(wrapper.find('i.ion-trash-a')).toHaveLength(0)
  })

  it('should trigger onDelete when trash icon clicked', function () {
    const onDelete = jest.fn()
    const comment = generateComments()
    comment.author = loggedUser
    const wrapper = shallow(<ArticleCommentCard onDelete={onDelete} comment={comment} />)

    wrapper.find('i.ion-trash-a').simulate('click')

    expect(onDelete).toBeCalledWith(comment.id)
  })
})
