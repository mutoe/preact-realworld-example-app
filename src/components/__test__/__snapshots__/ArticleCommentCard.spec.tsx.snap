// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Article comment card should render correctly 1`] = `
<div class="card">
  <div class="card-block">
    <p class="card-text">body.</p>
  </div>
  <div class="card-footer">
    <a href="/@Shirley Davis" class="comment-author">
      <img src="http://dummyimage.com/336x280" class="comment-author-img" />
    </a>Â 
    <a href="/@Shirley Davis" class="comment-author">Shirley Davis</a>
    <span class="date-posted">February 19</span>
    <span class="mod-options"></span>
  </div>
</div>
`;
